// angular-page-visibility/bundles/angular-page-visibility.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('angular-page-visibility', ['exports', '@angular/core', 'rxjs'], factory) :
    (global = global || self, factory(global['angular-page-visibility'] = {}, global.ng.core, global.rxjs));
}(this, (function (exports, core, rxjs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    (function (AngularPageVisibilityStateEnum) {
        AngularPageVisibilityStateEnum[AngularPageVisibilityStateEnum["VISIBLE"] = 0] = "VISIBLE";
        AngularPageVisibilityStateEnum[AngularPageVisibilityStateEnum["HIDDEN"] = 1] = "HIDDEN";
        AngularPageVisibilityStateEnum[AngularPageVisibilityStateEnum["PRERENDER"] = 2] = "PRERENDER";
        AngularPageVisibilityStateEnum[AngularPageVisibilityStateEnum["UNLOADED"] = 3] = "UNLOADED";
    })(exports.AngularPageVisibilityStateEnum || (exports.AngularPageVisibilityStateEnum = {}));

    var HiddenKeyConstant = /** @class */ (function () {
        function HiddenKeyConstant() {
        }
        HiddenKeyConstant.DEFAULT = "hidden";
        HiddenKeyConstant.MS = "msHidden";
        HiddenKeyConstant.WEB_KIT = "webkitHidden";
        return HiddenKeyConstant;
    }());
    var VisibilityStatusConstant = /** @class */ (function () {
        function VisibilityStatusConstant() {
        }
        VisibilityStatusConstant.VISIBLE = "visible";
        VisibilityStatusConstant.HIDDEN = "hidden";
        VisibilityStatusConstant.PRERENDER = "prerender";
        VisibilityStatusConstant.UNLOADED = "unloaded";
        return VisibilityStatusConstant;
    }());
    var AngularPageVisibilityService = /** @class */ (function () {
        function AngularPageVisibilityService() {
            this.onPageVisibleSource = new rxjs.Subject();
            this.onPageHiddenSource = new rxjs.Subject();
            this.onPagePrerenderSource = new rxjs.Subject();
            this.onPageUnloadedSource = new rxjs.Subject();
            this.onPageVisibilityChangeSource = new rxjs.Subject();
            this.$onPageVisible = this.onPageVisibleSource.asObservable();
            this.$onPageHidden = this.onPageHiddenSource.asObservable();
            this.$onPagePrerender = this.onPagePrerenderSource.asObservable();
            this.$onPageUnloaded = this.onPageUnloadedSource.asObservable();
            this.$onPageVisibilityChange = this.onPageVisibilityChangeSource.asObservable();
            this.addEventListenerVibilityChange();
        }
        AngularPageVisibilityService.prototype.isPageVisible = function () {
            return (VisibilityStatusConstant.VISIBLE === this.getVisibilityState() ||
                !this.isHidden());
        };
        AngularPageVisibilityService.prototype.isPageHidden = function () {
            return (VisibilityStatusConstant.HIDDEN === this.getVisibilityState() ||
                this.isHidden());
        };
        AngularPageVisibilityService.prototype.isPagePrerender = function () {
            return VisibilityStatusConstant.PRERENDER === this.getVisibilityState();
        };
        AngularPageVisibilityService.prototype.isPageUnloaded = function () {
            return VisibilityStatusConstant.UNLOADED === this.getVisibilityState();
        };
        AngularPageVisibilityService.prototype.isHidden = function () {
            return document[this.hidden];
        };
        AngularPageVisibilityService.prototype.getVisibilityState = function () {
            return document[this.visibilityState];
        };
        AngularPageVisibilityService.prototype.defineBrowserSupport = function () {
            if (typeof document[HiddenKeyConstant.DEFAULT] !== "undefined") {
                // Opera 12.10 and Firefox 18 and later support
                this.hidden = HiddenKeyConstant.DEFAULT;
                this.visibilityChange = "visibilitychange";
                this.visibilityState = "visibilityState";
            }
            else if (typeof document[HiddenKeyConstant.MS] !== "undefined") {
                this.hidden = HiddenKeyConstant.MS;
                this.visibilityChange = "msvisibilitychange";
                this.visibilityState = "msVisibilityState";
            }
            else if (typeof document[HiddenKeyConstant.WEB_KIT] !== "undefined") {
                this.hidden = HiddenKeyConstant.WEB_KIT;
                this.visibilityChange = "webkitvisibilitychange";
                this.visibilityState = "webkitVisibilityState";
            }
        };
        AngularPageVisibilityService.prototype.addEventListenerVibilityChange = function () {
            var _this = this;
            this.defineBrowserSupport();
            document.addEventListener(this.visibilityChange, function () {
                var vibilityState = _this.getVisibilityState();
                switch (vibilityState) {
                    case VisibilityStatusConstant.VISIBLE:
                        _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.VISIBLE);
                        _this.onPageVisibleSource.next();
                        break;
                    case VisibilityStatusConstant.HIDDEN:
                        _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.HIDDEN);
                        _this.onPageHiddenSource.next();
                        break;
                    case VisibilityStatusConstant.PRERENDER:
                        _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.PRERENDER);
                        _this.onPagePrerenderSource.next();
                        break;
                    case VisibilityStatusConstant.UNLOADED:
                        _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.UNLOADED);
                        _this.onPageUnloadedSource.next();
                        break;
                    default:
                        if (_this.isHidden()) {
                            _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.HIDDEN);
                            _this.onPageHiddenSource.next();
                        }
                        else {
                            _this.onPageVisibilityChangeSource.next(exports.AngularPageVisibilityStateEnum.VISIBLE);
                            _this.onPageVisibleSource.next();
                        }
                }
            }, false);
        };
        AngularPageVisibilityService.ɵprov = core.ɵɵdefineInjectable({ factory: function AngularPageVisibilityService_Factory() { return new AngularPageVisibilityService(); }, token: AngularPageVisibilityService, providedIn: "root" });
        AngularPageVisibilityService = __decorate([
            core.Injectable({
                providedIn: "root"
            }),
            __metadata("design:paramtypes", [])
        ], AngularPageVisibilityService);
        return AngularPageVisibilityService;
    }());

    var providers = [
        { provide: AngularPageVisibilityService, deps: [] },
    ];
    var injector = core.Injector.create({ providers: providers });
    var pageVisibilityService = injector.get(AngularPageVisibilityService);
    function OnPageVisibilityChange() {
        return function (target, propertyKey, descriptor) {
            var originalMethod = descriptor.value;
            var onPageHiddenSubscription;
            if (!target.ngOnInit) {
                buildNewMethodNgOnInit(target);
            }
            var originalNgOnInit = target.ngOnInit;
            target.ngOnInit = function () {
                var _this = this;
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageHiddenSubscription = pageVisibilityService.$onPageVisibilityChange.subscribe(function (visibilityState) {
                    return originalMethod.call(_this, [visibilityState]);
                });
                if (originalNgOnInit) {
                    originalNgOnInit.call(this, args);
                }
            };
            if (!target.ngOnDestroy) {
                buildNewMethodNgOnDestroy(target);
            }
            var originalNgOnDestroy = target.ngOnDestroy;
            target.ngOnDestroy = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageHiddenSubscription.unsubscribe();
                if (originalNgOnDestroy) {
                    originalNgOnDestroy.call(this, args);
                }
            };
        };
    }
    function OnPageHidden() {
        return function (target, propertyKey, descriptor) {
            var originalMethod = descriptor.value;
            var onPageHiddenSubscription;
            if (!target.ngOnInit) {
                buildNewMethodNgOnInit(target);
            }
            var originalNgOnInit = target.ngOnInit;
            target.ngOnInit = function () {
                var _this = this;
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageHiddenSubscription = pageVisibilityService.$onPageHidden.subscribe(function () { return originalMethod.call(_this); });
                if (originalNgOnInit) {
                    originalNgOnInit.call(this, args);
                }
            };
            if (!target.ngOnDestroy) {
                buildNewMethodNgOnDestroy(target);
            }
            var originalNgOnDestroy = target.ngOnDestroy;
            target.ngOnDestroy = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageHiddenSubscription.unsubscribe();
                if (originalNgOnDestroy) {
                    originalNgOnDestroy.call(this, args);
                }
            };
        };
    }
    function OnPageVisible() {
        return function (target, propertyKey, descriptor) {
            var originalMethod = descriptor.value;
            if (!target.ngOnInit) {
                buildNewMethodNgOnInit(target);
            }
            var originalNgOnInit = target.ngOnInit;
            var onPageVisibleSubscription;
            target.ngOnInit = function () {
                var _this = this;
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageVisibleSubscription = pageVisibilityService.$onPageVisible.subscribe(function () { return originalMethod.call(_this); });
                if (originalNgOnInit) {
                    originalNgOnInit.call(this, args);
                }
            };
            if (!target.ngOnDestroy) {
                buildNewMethodNgOnDestroy(target);
            }
            var originalNgOnDestroy = target.ngOnDestroy;
            target.ngOnDestroy = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageVisibleSubscription.unsubscribe();
                if (originalNgOnDestroy) {
                    originalNgOnDestroy.call(this, args);
                }
            };
        };
    }
    function OnPagePrerender() {
        return function (target, propertyKey, descriptor) {
            var originalMethod = descriptor.value;
            var onPagePrerenderSubscription;
            if (!target.ngOnInit) {
                buildNewMethodNgOnInit(target);
            }
            var originalNgOnInit = target.ngOnInit;
            target.ngOnInit = function () {
                var _this = this;
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPagePrerenderSubscription = pageVisibilityService.$onPagePrerender.subscribe(function () { return originalMethod.call(_this); });
                if (originalNgOnInit) {
                    originalNgOnInit.call(this, args);
                }
            };
            if (!target.ngOnDestroy) {
                buildNewMethodNgOnDestroy(target);
            }
            var originalNgOnDestroy = target.ngOnDestroy;
            target.ngOnDestroy = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPagePrerenderSubscription.unsubscribe();
                if (originalNgOnDestroy) {
                    originalNgOnDestroy.call(this, args);
                }
            };
        };
    }
    function OnPageUnloaded() {
        return function (target, propertyKey, descriptor) {
            var originalMethod = descriptor.value;
            var onPageUnloadedSubscription;
            if (!target.ngOnInit) {
                buildNewMethodNgOnInit(target);
            }
            var originalNgOnInit = target.ngOnInit;
            target.ngOnInit = function () {
                var _this = this;
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageUnloadedSubscription = pageVisibilityService.$onPageUnloaded.subscribe(function () { return originalMethod.call(_this); });
                if (originalNgOnInit) {
                    originalNgOnInit.call(this, args);
                }
            };
            if (!target.ngOnDestroy) {
                buildNewMethodNgOnDestroy(target);
            }
            var originalNgOnDestroy = target.ngOnDestroy;
            target.ngOnDestroy = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                onPageUnloadedSubscription.unsubscribe();
                if (originalNgOnDestroy) {
                    originalNgOnDestroy.call(this, args);
                }
            };
        };
    }
    function buildNewMethodNgOnInit(target) {
        newMethod(target, 'ngOnInit');
    }
    function buildNewMethodNgOnDestroy(target) {
        newMethod(target, 'ngOnDestroy');
    }
    function newMethod(target, name) {
        Object.defineProperty(target, name, {
            value: function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
            },
            writable: true,
        });
    }

    var AngularPageVisibilityModule = /** @class */ (function () {
        function AngularPageVisibilityModule() {
        }
        AngularPageVisibilityModule = __decorate([
            core.NgModule({
                imports: [],
                declarations: [],
                providers: [AngularPageVisibilityService],
                exports: []
            })
        ], AngularPageVisibilityModule);
        return AngularPageVisibilityModule;
    }());

    exports.AngularPageVisibilityModule = AngularPageVisibilityModule;
    exports.AngularPageVisibilityService = AngularPageVisibilityService;
    exports.OnPageHidden = OnPageHidden;
    exports.OnPagePrerender = OnPagePrerender;
    exports.OnPageUnloaded = OnPageUnloaded;
    exports.OnPageVisibilityChange = OnPageVisibilityChange;
    exports.OnPageVisible = OnPageVisible;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
